packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.kt

```diff
@@ -0,0 +1,1377 @@
+          return requestBuilder.build()
+        }
+        else -> {
+          promise.reject("ERR_FILESYSTEM_INVALID_UPLOAD_TYPE", String.format("Invalid upload type: %s.", options["uploadType"]), null)
+          return null
+        }
+      }
+    } catch (e: Exception) {
+      e.message?.let { Log.e(TAG, it) }
+      promise.reject(e)
+    }
+    return null
+  }
+
+  @ExpoMethod
+  fun uploadAsync(url: String, fileUriString: String, options: Map<String, Any>, promise: Promise) {
+    val request = createUploadRequest(
+      url, fileUriString, options, promise,
+      object : RequestBodyDecorator {
+        override fun decorate(requestBody: RequestBody): RequestBody {
+          return requestBody
+        }
+      }
+    ) ?: return
+
+    okHttpClient?.let {
+      it.newCall(request).enqueue(object : Callback {
+        override fun onFailure(call: Call, e: IOException) {
+          Log.e(TAG, e.message.toString())
+          promise.reject(e)
+        }
+
+        override fun onResponse(call: Call, response: Response) {
+          val result = Bundle().apply {
+            putString("body", response.body()?.string())
+            putInt("status", response.code())
+            putBundle("headers", translateHeaders(response.headers()))
+          }
+          response.close()
+          promise.resolve(result)
+        }
+      })
+    } ?: run {
+      promise.reject(NullPointerException("okHttpClient is null"))
+    }
+  }
+
+  @ExpoMethod
+  fun uploadTaskStartAsync(url: String, fileUriString: String, uuid: String, options: Map<String, Any>, promise: Promise) {
+    val progressListener: CountingRequestListener = object : CountingRequestListener {
+      private var mLastUpdate: Long = -1
+      override fun onProgress(bytesWritten: Long, contentLength: Long) {
+        val eventEmitter: EventEmitter by moduleRegistry()
+        val uploadProgress = Bundle()
+        val uploadProgressData = Bundle()
+        val currentTime = System.currentTimeMillis()
+
+        // Throttle events. Sending too many events will block the JS event loop.
+        // Make sure to send the last event when we're at 100%.
+        if (currentTime > mLastUpdate + MIN_EVENT_DT_MS || bytesWritten == contentLength) {
+          mLastUpdate = currentTime
+          uploadProgressData.putDouble("totalByteSent", bytesWritten.toDouble())
+          uploadProgressData.putDouble("totalBytesExpectedToSend", contentLength.toDouble())
+          uploadProgress.putString("uuid", uuid)
+          uploadProgress.putBundle("data", uploadProgressData)
+          eventEmitter.emit(EXUploadProgressEventName, uploadProgress)
+        }
+      }
+    }
+    val request = createUploadRequest(
+      url,
+      fileUriString,
+      options,
+      promise,
+      object : RequestBodyDecorator {
+        override fun decorate(requestBody: RequestBody): RequestBody {
+          return CountingRequestBody(requestBody, progressListener)
+        }
+      }
```

Comment:
If `RequestBodyDecorator` is a functional interface (Java interface with a single function) or Kotlin `fun interface`, then this could be converted to a lambda:
```suggestion
      { requestBody -> CountingRequestBody(requestBody, progressListener) }
```